src/nodes/mod.rs

```diff
@@ -61,13 +62,41 @@ impl From<io::Error> for KernelError {
     }
 }
 
+impl From<nix::Error> for KernelError {
+    fn from(e: nix::Error) -> Self {
+        match e {
+            nix::Error::Sys(errno) => KernelError::from_errno(errno),
+            _ => {
+                warn!("Got nix::Error without an errno; propagating as EIO: {}", e);
+                KernelError::from_errno(Errno::EIO)
+            }
+        }
+    }
+}
+
 /// Generic result type for of all node operations.
 pub type NodeResult<T> = Result<T, KernelError>;
 
 /// Abstract representation of an open file handle.
 pub trait Handle {
     /// Reads `size` bytes from the open file starting at `offset`.
-    fn read(&self, offset: i64, size: u32) -> NodeResult<Vec<u8>>;
+    fn read(&self, _offset: i64, _size: u32) -> NodeResult<Vec<u8>> {
+        panic!("Not implemented")
+    }
+
+    /// Reads all directory entries into the given reply object.
+    ///
+    /// While this takes a `fuse::ReplyDirectory` object as a parameter for efficiency reasons, it
+    /// is the responsibility of the caller to invoke `reply.ok()` and `reply.error()` on the same
+    /// reply object.  This is for consistency with the handling of any errors returned by this and
+    /// other functions.
+    ///
+    /// `_ids` and `_cache` are the file system-wide bookkeeping objects needed to instantiate new
+    /// nodes, used when readdir discovers an underlying node that was not yet known.
+    fn readdir(&self, _ids: &IdGenerator, _cache: &Cache, _reply: &mut fuse::ReplyDirectory)
+        -> NodeResult<()> {
+        panic!("Not implemented");
```

Comment:
Return ENOTDIR by default.

Reply:
I see what you are trying to guard against, but I don't think the kernel should ever invoke this behavior. I've added an integration test instead.

Reply:
... and even if it does, it seems that the FUSE libraries are guarding against it based on some inconsistencies I have observed in return error codes between macOS/Linux and Go/Rust. So I don't think I should have code that is never executed -- and if it _is_ ever executed, I'd like to know why.